subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java

```diff
@@ -175,8 +178,20 @@ public Configuration createWithRole(String name, ConfigurationRole role, boolean
     }
 
     private ConfigurationInternal doCreate(String name, ConfigurationRole role, boolean lockUsage) {
+        validateNameIsAllowed(name);
         DefaultConfiguration configuration = newConfiguration(name, this, rootComponentMetadataBuilder, role, lockUsage);
         configuration.addMutationValidator(rootComponentMetadataBuilder.getValidator());
         return configuration;
     }
+
+    private void validateNameIsAllowed(String name) {
+        if (RESERVED_NAMES_FOR_DETACHED_CONFS.matcher(name).matches()) {
+            String message = String.format("Creating a configuration named '%s' (which begins with 'detachedConfiguration' and is optionally followed by a number)", name);
+            DeprecationLogger.deprecateAction(message)
+                    .withAdvice("Use a different name for this configuration.")
+                    .willBeRemovedInGradle9()
+                    .withUpgradeGuideSection(8, "reserved_configuration_names")
+                    .nagUser();
```

Comment:
I would be a little more straightforward:

```suggestion
            DeprecationLogger.deprecateAction("Creating a configuration with a name that starts with 'detachedConfiguration'")
                    .withAdvice(String.format("Use a different name for the configuration '%s'.", name))
                    .willBeRemovedInGradle9()
                    .withUpgradeGuideSection(8, "reserved_configuration_names")
                    .nagUser();
```

This should look something like:
> Creating a configuration with a name that starts with 'detachedConfiguration' has been deprecated. This is scheduled to be removed in Gradle 9.0. Use a different name for the configuration '$name'.